src/Routing.purs

```diff
@@ -1,67 +1,92 @@
-
-
-hashes :: forall e. (String -> String -> Eff e Unit) -> Eff e Unit
-hashes cb =
-  hashChanged $ \old new -> do
-    cb (dropHash old) (dropHash new)
-  where dropHash h =
-          case R.regex "^[^#]*#" RF.noFlags of
-            Right regX -> R.replace regX "" h
-            Left _     -> h
-
-
--- | Stream of hash changed, callback called when new hash can be matched
--- | First argument of callback is `Just a` when old hash can be matched
--- | and `Nothing` when it can't.
-matches :: forall e a. Match a -> (Maybe a -> a -> Eff e Unit) -> Eff e Unit
-matches = matches' decodeURIComponent
-
-matches' :: forall e a. (String -> String) ->
-            Match a -> (Maybe a -> a -> Eff e Unit) -> Eff e Unit
-matches' decoder routing cb = hashes $ \old new ->
-  let mr = matchWith decoder routing
-      fst = either (const Nothing) Just $ mr old
-  in either (const $ pure unit) (cb fst) $ mr new
+-- | Folds effectfully over hash changes given a callback and an initial hash.
+-- | The provided String is the hash portion of the `Location` with the '#'
+-- | prefix stripped. Returns an effect which will remove the listener.
+foldHashes
+  :: forall eff a
+   . (a -> String -> Eff (RoutingEffects eff) a)
+  -> (String -> Eff (RoutingEffects eff) a)
+  -> Eff (RoutingEffects eff) (Eff (RoutingEffects eff) Unit)
+foldHashes cb init = do
+  ref <- newRef =<< init =<< getHash
+  win <- windowToEventTarget <$> window
+  let listener = eventListener \_ -> writeRef ref =<< join (cb <$> readRef ref <*> getHash)
+  addEventListener hashchange listener false win
+  pure $ removeEventListener hashchange listener false win
 
-matchesAff' :: forall e a. (String -> String) ->
-               Match a -> Aff e (Tuple (Maybe a) a)
-matchesAff' decoder routing =
-  makeAff \k -> nonCanceler <$
-    matches' decoder routing \old new ->
-      k $ Right $ Tuple old new
+-- | Runs the callback on every hash change providing the previous hash and the
+-- | latest hash. The provided String is the hash portion of the `Location` with
+-- | the '#' prefix stripped. Returns an effect which will remove the listener.
+hashes
+  :: forall eff
+   . (Maybe String -> String -> Eff (RoutingEffects eff) Unit)
+  -> Eff (RoutingEffects eff) (Eff (RoutingEffects eff) Unit)
+hashes = matchesWith Just
 
-matchesAff :: forall e a. Match a -> Aff e (Tuple (Maybe a) a)
-matchesAff = matchesAff' decodeURIComponent
+-- | Runs the callback on every hash change using a given `Match` parser to
+-- | extract a route from the hash. If a hash fails to parse, it is ignored.
+-- | To avoid dropping hashes, provide a fallback alternative in your parser.
+-- | Returns an effect which will remove the listener.
+matches
+  :: forall eff a
+   . Match a
+  -> (Maybe a -> a -> Eff (RoutingEffects eff) Unit)
+  -> Eff (RoutingEffects eff) (Eff (RoutingEffects eff) Unit)
+matches matcher = matchesWith (matchWith decodeURIComponent matcher)
```

Comment:
it's same as `matches = matchesWith <<< match`

Reply:
Good catch